### ISSUE TITLE
Cannot get Jooby 2 Running on Tomcat 9

### ISSUE THREAD

**Issue by nikolaosinlight** [jooby-project/jooby#2446]
Hi,

I have been building a Jooby 2 App (Website/Portal) and testing with Jetty/IntelliJ and we are days from Launch.

Now, I am trying to install it on Tomcat.  I was able to:
1. Crate a WAR for the project 
2. Create a Servlet Wrapper (via web.xml) that kicks off Jooby.main()
3. Worked through an issue with Assets needing to be pulled from class path vs. absolute path
4. And now with Tomcat firing up Netty complains that 8080 is taken.

What do I need to do to make Netty not come up?  I see there are many Jooby2 overloaded methods... just not sure what to do?  i.e. should I not define any server parameters?  Any way to specify NONE?

We have been deploying production Java Applications on Tomcat for years and have the complete configuration ready to with Virtual Hosts, Multiple SSL certificates via SNI, etc.... and want to keep our deployments on Tomcat.

How do I get Jooby 2 to run on Tomcat???

--Nikolaos

**Comment by nikolaosinlight** [comment]
Hi Edgar,

Wow.  With the great design and modular aspects of Jooby 2 I am very surprised that one cannot turn off the web server. 

As an alternative... if we used Jetty (closest thing to Tomcat) and we setup virtual hosts could we run multiple Jooby 2 webapps (WAR files) inside a single Jetty server?  e.g. I have:

https://www.example1.com (port 443) going to /webapp1
https://www.example2.com (port 443) going to /webapp2

Both running within the same Jetty web container.

Or do we need to run multiple copies of Jooby 2 each on their own port (I sincerely hope not as this would be awful as we are deploying multiple sites on the same platform).

--Nikolaos

**Comment by jknack** [comment]
jooby 1.x used to support war deployments, but almost no one ever used it. It doesn't worth maintaining it, but also there were a couple of limitations.

> 1. Would offering Tomcat (via WAR) support be on table for the next release of Jooby or are you against this idea outright?

Not from me.

> 2. Would you be willing if I tried to customize Jooby to have this support i.e. I already have Jooby running from a WAR file in Tomcat on a virtual host (trivial really) it is just that it is trying to launch Netty (if I could severe the link to launching Netty wouldn't the JARs just work within the JVM as long as the classloader can find the required dependencies, etc.)?

You need to:

1. Write a servlet to bootstrap Jooby (as you already did)
2. Implement io.jooby.Context using servlet API (HttpServletRequest, HttpServletResponse)

It is a lot of work.

> But single instance per site is really going to be impractical to stick with Jooby as one cannot host > 1 site without:

Jooby offers alternatives for this, far more simple and efficient:

1. https://jooby.io/#router-composing
2. https://jooby.io/#router-dynamic-routing

Example with a domain:

```java
{
   mount(ctx -> ctx.getHost().equals("foo.com"), new FooApp());
   mount(ctx -> ctx.getHost().equals("bar.com"), new BarApp());
 }
```

**Comment by jknack** [comment]
> Lastly, you say "Implement io.jooby.Context" is a lot of work... would you have any pointers if the above isn't that great?

Yea, you should look at jooby-jetty it is close to pure servlet deployment.

**Comment by nikolaosinlight** [comment]
So I looked at jooby-jetty and see what you mean about a lot of code to support.

Can you let me know if 1, 2, and 3 make sense (in my last post).  Any other tips or pointers to handle multiple domains?

Much Appreciated!

--Nikolaos

**Comment by nikolaosinlight** [comment]
@agentgt I specialize in IDM (ForgeRock) and currently do PS where we deploy OpenAM, OpenDJ and OpenIDM in AWS pods (K8 and docker) for Fortune 100/500 clients but guess what the OpenAM deployed in a pod is running a tomcat server and the enterprise grade application is a WAR that is deployed in the tomcat container :-)

K8 / Docker does not preclude WAR files.  Just because uber/fat JARs have become fashionable with micro services does not mean that WAR files are dead or that Java application servers like Tomcat, JBoss, WebLogic, WebSphere are dead.  In fact, last I checked Tomcat had 50% market share and Jetty was just 9% and moreover Tomcat is just growing in support over the years as it is being used in place of those other heavy weight Java / J2EE servers.

BTW K8 / Docker has its horror stories as well... it isn't all peaches... just google K8 OOM Killer and you will find dozens of articles of people pulling their heads out b/c Java 8 and even Java 11 (although it was supposed to be fixed) have JVMs that grow to the size of the container and kill the pod by choking the OS.

With that said, I still don't think a web container should be single application but that is just IMHO and as I would prefer to stick to Jooby may end up doing phase II of our servers on K8 / Docker with Jooby PoC in the next month or so....

ASIDE:  WAR files elegantly package any application... uber/fat JARs are horrible the way everything gets tossed into a compressed archive with so many warnings about duplicates / overwriting of class files your head is spinning... and that is if you are lucky some newer needed version of something isn't overwritten by older one... definitely going backwards on this front as well I am afraid.  Lastly, nice to be able to launch a JAR file from command line on your laptop/workstation but what is needed is a server to run the thing so you end up writing/finding a script wrapper around the thing when Java servers run as services and just have everything you need to simply run your code.  Like I said... going backwards IMHO....

--Nikolaos

**Comment by agentgt** [comment]
@nikolaosinlight I only assumed you wanted to package multiple services with varying context paths (the path after the host) into a single deployment and thus gave you the idea of docker.

I'm kind of curious why you would want to use Jooby if you are already using the full JEE stack (e.g. servlet container).

And I didn't mean to denigrate WARs (I have more experience than you probably think dealing with them).

I'm sure lots of people still use servlet containers and the original JEE stack tech just like most companies are stuck on Java 8 but new development is not focused on WAR like deployments and you can clearly see that with things like Micronaut, Quarkus, Microprofile, Helidon, Spring Boot Native etc... speaking of native its hard to Graal native-image a WAR.

I'm sure Tomcat's growth has improved (albeit I find some of your stats dubious w/o sources) but I'm also fairly sure that is because Spring Boot ships it as a default. I'm not sure how much longer that will happen (tomcat being the default). Probably depends on Loom and Spring Native future.

**Comment by nikolaosinlight** [comment]
@agentgt I didn't mean any disrespect on skills.  My point is that every technology has its PROs and CONs and yes K8 is extremely enticing but I would not agree that "_Docker and friends basically killed servlet container deployments (aka WAR)._". There is a saying that the grass is greener on the other side... but I often add... sometimes it is burnt ;-).  

**As for why Jooby?**

After doing Java since its inception with 1.1.x from Fat Clients, Video Servers, large Web Sites / Portals... Java does not make small/medium sized software development easy.  So a number of years ago I tried the php route with FatFree (f3) (awesome micro framework very much like Jooby) but found php to be heavy and not very scalable.  A few years ago I came across Jooby 1.0 and basically found what is akin to FatFree in Java and fully embraced it for all our in-house projects. Jooby 1.0 checked off all our boxes and IIRC supported WARs / Tomcat to boot.  I didn't realize until we were close to the tail end of our project that Jooby 2.0 had dropped WAR functionality and completely get why it is so.

Jooby is absolutely amazing when it comes to its clean architecture and module approach (even if 2.0 cut down on some) and essentially makes building small to medium sized apps that can scale really easy.

Quite frankly we are not married to our deployment technology (be it WAR / Tomcat) but HUGE on Jooby so at this point we will likely launch Jooby Apps using Stork for MVP 1.0 but... in a large part Thanks to You and this conversation... will quickly look to using K8 / Docker especially as Ubuntu 20.04 LTS promotes MicroK8s (every time you login... LOL) and will go from there.

Thanks Again for your perspective.  Much appreciated and Kind Regards!
